src/Development/IDE/Core/Shake.hs

```diff
@@ -222,13 +236,25 @@ type IdeRule k v =
   , NFData v
   )
 
+-- | A live Shake session with the ability to enqueue Actions for running.
+--   Keeps the 'ShakeDatabase' open, so at most one 'ShakeSession' per database.
+data ShakeSession = ShakeSession
+  { cancelShakeSession :: !(IO [Action ()])
+    -- ^ Closes the Shake session and returns the pending user actions
+  , runInShakeSession  :: !(forall a . Action a -> IO (IO a))
+    -- ^ Enqueue a user action in the Shake session.
+  }
+
+emptyShakeSession :: ShakeSession
+emptyShakeSession = ShakeSession (pure []) (\_ -> error "emptyShakeSession")
+
 -- | A Shake database plus persistent store. Can be thought of as storing
 --   mappings from @(FilePath, k)@ to @RuleResult k@.
 data IdeState = IdeState
-    {shakeDb :: ShakeDatabase
-    ,shakeAbort :: MVar (IO ()) -- close whoever was running last
-    ,shakeClose :: IO ()
-    ,shakeExtras :: ShakeExtras
+    {shakeDb         :: ShakeDatabase
+    ,shakeSession    :: MVar ShakeSession
```

Comment:
Why is `shakeSession` wrapped in an `MVar` if you are already using a `TQueue`?

Reply:
The `TQueue` cannot be closed, so the `MVar` is needed to ensure that no more actions are added while the session is being restarted. It's likely that this can be simplified further, this is mostly an iteration on top of the previous design.